### DISCUSSION TITLE
Visibility of the Stream adaptor structures in tokio-stream

### DISCUSSION THREAD

**Discussion by sharpened-nacho** [tokio-rs/tokio#6655] [General] [answered]
The `StreamExt` trait provides the helpful method `take`, `map`, and the like on `Stream`s. But the structures themselves (`Take`, `Map`, etc.) are not visible outside the crate, as they are not `pub use` in `stream_ext.rs`.
Is there a reason for that?

For context, I am playing around with tonic, and need to name the type of the `Stream` I use to implement a Server Streaming RPC. So :
```rust
impl MyServiceTrait for MyServiceStructure {
    type RpcResponseType = Take<MyStreamType>;

    async fn my_fn(&self, request: ...) -> Result<Response<RpcResponseType>, Status> { ... }
}
```
and I can't import `Take`.

**Comment by Darksonn** [comment]
We usually do this kind of thing so that we can later use `impl Trait` as the return value in the extension trait without a breaking change. But in the case of tokio-stream this seems less important, as we can just make a breaking change.

**Reply by sharpened-nacho** [reply]
Sorry, I don't think I understand the explanation :/ Are you referring to changing, say, `fn take(self, n: usize) -> Take<Self>` to `fn take(self, n: usize)-> impl Stream` ? In which case I don't get why the visibility of Take has any influence in that, which makes me think I just didn't get it.
As for the 2nd sentence, does it imply some change will be made?

**Reply by Darksonn** [reply, marked as answer]
If the type is visible, then you can write code that would not compile if the type stops being `Take<Self>`. Changing it to return `impl Stream` would count as that.

As for whether a change will be made, well, I could probably be convinced to accept a change for this. But I will not make the change myself.

**Reply by sharpened-nacho** [reply]
Thank you for clarifying.
I understand the comfort of having the option to change the type, but I must admit I don't see a context for it - I have no experience in maintaining libs though.

I can do the change myself then. I would have asked to - it's a great first issue and PR as it's simple. I will study the contribution guide, make sure I understand everything, and come back in issues.
Thanks for the responses!

**Reply by Darksonn** [reply]
It's more important for `AsyncReadExt` where we want to change `fn next() -> Read` to `async fn next() -> usize` in the future.
